beacon-chain/p2p/sender.go

```diff
@@ -8,28 +8,44 @@ import (
 	"github.com/libp2p/go-libp2p-core/network"
 	"github.com/libp2p/go-libp2p-core/peer"
 	"github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/prysmaticlabs/prysm/shared/traceutil"
+	"go.opencensus.io/trace"
 )
 
 // Send a message to a specific peer. The returned stream may be used for reading, but has been
 // closed for writing.
 func (s *Service) Send(ctx context.Context, message interface{}, pid peer.ID) (network.Stream, error) {
+	ctx, span := trace.StartSpan(ctx, "p2p.Send")
+	defer span.End()
 	topic := RPCTypeMapping[reflect.TypeOf(message)] + s.Encoding().ProtocolSuffix()
+	span.AddAttributes(trace.StringAttribute("topic", topic))
 
 	// TTFB_TIME (5s) + RESP_TIMEOUT (10s).
-	ctx, cancel := context.WithTimeout(ctx, 15*time.Second)
+	const deadline = 15*time.Second
+	ctx, cancel := context.WithTimeout(ctx, deadline)
 	defer cancel()
 
 	stream, err := s.host.NewStream(ctx, pid, protocol.ID(topic))
 	if err != nil {
+		traceutil.AnnotateError(span, err)
+		return nil, err
+	}
+	if err := stream.SetReadDeadline(time.Now().Add(deadline)); err != nil {
```

Comment:
maybe all stream deadlines related methods should be in `p2p`, what do you think ? It doesnt seem related to sync specifically , its more of a stream managment issue, so maybe the p2p package would be the best place for it to live in. We can shift `deadlines.go` into `p2p`